p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/FtpTransportProtocolHandler.java

```diff
@@ -0,0 +1,200 @@
+            return -1;
+        }
+
+        final FTPFile file = client.mdtmFile(uri.getPath());
+
+        if (file == null) {
+            throw new IOException("Could not find file: " + uri);
+        }
+
+        return file.getTimestampInstant().toEpochMilli();
+    }
+
+    @Override
+    public File getFile(final URI uri) throws IOException {
+        final File localFile = getLocalFile(uri);
+
+        if (cacheConfig.isOffline()) {
+            if (!localFile.isFile()) {
+                throw new IOException("Maven is offline and the requested file does not exist locally: " + uri);
+            }
+
+            return localFile;
+        }
+
+        final FTPClient client = getClient(uri);
+        final String remotePath = uri.getPath();
+        final FTPFile remoteFile = getRemoteFile(client, remotePath);
+        final boolean isRemoteMissing = remoteFile == null;
+
+        if (localFile.isFile() && (isRemoteMissing || !mustRefresh(localFile, remoteFile))) {
+            return localFile;
+        }
+
+        if (isRemoteMissing) {
+            logger.warn("Could not find file: " + uri);
+            return null;
+        }
+
+        final File parent = FileUtils.createParentDirectories(localFile);
+        final File tempFile = File.createTempFile("download", ".tmp", parent);
+
+        try (final OutputStream os = new FileOutputStream(tempFile)) {
+            if (!client.retrieveFile(remotePath, os)) {
+                final var message = client.getReplyString();
+                throw new IOException(String.format("Error retrieving file: %s. Message: %s", remotePath, message));
+            }
+        } catch (final IOException e) {
+            tempFile.delete();
+            throw e;
+        }
+
+        if (localFile.isFile()) {
+            FileUtils.forceDelete(localFile);
+        }
+
+        FileUtils.moveFile(tempFile, localFile);
+        localFile.setLastModified(remoteFile.getTimestampInstant().toEpochMilli());
+        return localFile;
+    }
+
+    public void dispose() {
+        for (final Map.Entry<String, FTPClient> entry : clients.entrySet()) {
+            try {
+                entry.getValue().logout();
+                entry.getValue().disconnect();
+            } catch (final FTPConnectionClosedException e) {
+                // Connection already closed by the host
+            } catch (final IOException e) {
+                logger.error("Error disconnecting from host: " + entry.getKey(), e);
+            }
+        }
+    }
+
+    protected FTPClient createFtpClient() {
+        final FTPClient client = new FTPClient();
+        client.setControlKeepAliveTimeout(Duration.ofSeconds(15));
+        client.setAutodetectUTF8(true);
+        client.setBufferSize(8192);
+        return client;
+    }
+
+    private FTPClient getClient(final URI uri) throws IOException {
+        final String host = uri.getHost();
+        final int port = uri.getPort() < 0 ? 21 : uri.getPort();
```

Comment:
Can we extract the number 21 into a constant, probably even with a systemproperty using Integer.getInteger("...", 21)?

Reply:
The `tycho.p2.transport.ftp.port` property can now change the default FTP port.